File: application/app/src/test/java/com/yahyeet/boardbook/FirebaseMockTest.java
package com.yahyeet.boardbook;

import com.yahyeet.boardbook.model.Boardbook;
import com.yahyeet.boardbook.model.entity.Game;
import com.yahyeet.boardbook.model.entity.Match;
import com.yahyeet.boardbook.model.entity.User;
import com.yahyeet.boardbook.model.handler.AuthHandler;
import com.yahyeet.boardbook.model.handler.GameHandler;
import com.yahyeet.boardbook.model.handler.MatchHandler;
import com.yahyeet.boardbook.model.handler.UserHandler;
import com.yahyeet.boardbook.mock.repository.MockGameRoleRepository;
import com.yahyeet.boardbook.mock.repository.MockGameTeamRepository;
import com.yahyeet.boardbook.mock.repository.MockMatchPlayerRepository;
import com.yahyeet.boardbook.mock.service.MockAuthService;
import com.yahyeet.boardbook.mock.repository.MockGameRepository;
import com.yahyeet.boardbook.mock.repository.MockMatchRepository;
import com.yahyeet.boardbook.mock.repository.MockUserRepository;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.concurrent.ExecutionException;

import static org.junit.Assert.*;

public class FirebaseMockTest {


    MockAuthService authService;

    MockUserRepository userRepository;
    MockGameRepository gameRepository;
    MockMatchRepository matchRepository;
    MockGameTeamRepository teamRepository;
    MockMatchPlayerRepository playerRepository;
    MockGameRoleRepository roleRepository;

    Boardbook boardbook;


    @Before
    public void initTests() {
        authService = new MockAuthService(new ArrayList<>());

        userRepository = new MockUserRepository();
        gameRepository = new MockGameRepository();
        matchRepository = new MockMatchRepository();
        teamRepository = new MockGameTeamRepository();
        playerRepository = new MockMatchPlayerRepository();
        roleRepository = new MockGameRoleRepository();


        boardbook = new Boardbook(
                new AuthHandler(authService),
                new UserHandler(userRepository, matchRepository, roleRepository, teamRepository,gameRepository, playerRepository),
                new GameHandler(gameRepository, roleRepository, teamRepository, matchRepository),
                new MatchHandler(matchRepository, playerRepository, gameRepository, roleRepository, teamRepository, userRepository)
        );

    }


    // This test works in its completableFutureness, write tests like this
    @Test
    public void testLogin() throws ExecutionException, InterruptedException {
        User user = new User("TestUser");

        User friend = new User("Friendly");
        Game game = new Game("Avalon", "Cool game", 1, 2, 3);
        Match match = new Match(game);

        user.getFriends().add(friend);
        user.getMatches().add(match);


        MockAuthService.AuthenticationUser authUser = new MockAuthService.AuthenticationUser("email@email.test", "wordpass", user);

        boardbook.getAuthHandler().signup(authUser.getEmail(), authUser.getPassword(), authUser.getUser().getName()).thenAccept(user1 -> {
            assertEquals(user.getName(), boardbook.getAuthHandler().getLoggedInUser().getName());
        }).get();

    }

}

File: application/app/src/test/java/com/yahyeet/boardbook/FirebaseMockTest.kt
package com.yahyeet.boardbook

import com.yahyeet.boardbook.model.Boardbook
import com.yahyeet.boardbook.model.entity.Game
import com.yahyeet.boardbook.model.entity.Match
import com.yahyeet.boardbook.model.entity.User
import com.yahyeet.boardbook.model.handler.AuthHandler
import com.yahyeet.boardbook.model.handler.GameHandler
import com.yahyeet.boardbook.model.handler.MatchHandler
import com.yahyeet.boardbook.model.handler.UserHandler
import com.yahyeet.boardbook.model.mock.repository.MockGameRoleRepository
import com.yahyeet.boardbook.model.mock.repository.MockGameTeamRepository
import com.yahyeet.boardbook.model.mock.repository.MockMatchPlayerRepository
import com.yahyeet.boardbook.model.mock.service.MockAuthService
import com.yahyeet.boardbook.model.mock.repository.MockGameRepository
import com.yahyeet.boardbook.model.mock.repository.MockMatchRepository
import com.yahyeet.boardbook.model.mock.repository.MockUserRepository

import org.junit.Before
import org.junit.Test

import java.util.ArrayList
import java.util.concurrent.ExecutionException

import org.junit.Assert.*

class FirebaseMockTest {


    internal lateinit var authService: MockAuthService

    internal lateinit var userRepository: MockUserRepository
    internal lateinit var gameRepository: MockGameRepository
    internal lateinit var matchRepository: MockMatchRepository
    internal lateinit var teamRepository: MockGameTeamRepository
    internal lateinit var playerRepository: MockMatchPlayerRepository
    internal lateinit var roleRepository: MockGameRoleRepository

    internal lateinit var boardbook: Boardbook


    @Before
    fun initTests() {
        authService = MockAuthService(ArrayList())

        userRepository = MockUserRepository()
        gameRepository = MockGameRepository()
        matchRepository = MockMatchRepository()
        teamRepository = MockGameTeamRepository()
        playerRepository = MockMatchPlayerRepository()
        roleRepository = MockGameRoleRepository()


        boardbook = Boardbook(
                AuthHandler(authService),
                UserHandler(userRepository, matchRepository, roleRepository, teamRepository, gameRepository, playerRepository),
                GameHandler(gameRepository, roleRepository, teamRepository, matchRepository),
                MatchHandler(matchRepository, playerRepository, gameRepository, roleRepository, teamRepository, userRepository)
        )

    }


    // This test works in its completableFutureness, write tests like this
    @Test
    @Throws(ExecutionException::class, InterruptedException::class)
    fun testLogin() {
        val user = User("TestUser")

        val friend = User("Friendly")
        val game = Game("Avalon", "Cool game", 1, 2, 3)
        val match = Match(game)

        user.friends.add(friend)
        user.matches.add(match)


        val authUser = MockAuthService.AuthenticationUser("email@email.test", "wordpass", user)

        authUser.user.name?.let {
            boardbook.authHandler.signup(authUser.email, authUser.password, it).thenAccept { user1 ->
            assertEquals(user.name, boardbook.authHandler.loggedInUser?.name)
            //TODO: Add more assert equals, id?
        }.get()
        }

    }

}
